Query Label Names and Values from blocks store

gouthamve opened the pull request:
Signed-off-by: Goutham Veeramachaneni <gouthamve@gmail.com>

<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with master
-->

**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #2681

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pracucci commented on integration/querier_test.go:
```diff
@@ -590,7 +590,7 @@ func testMetadataQueriesWithBlocksStorage(
 					exp = append(exp, model.LabelValue(val))
 				}
-				require.ElementsMatch(t, exp, labelsRes)
+				require.Equal(t, exp, labelsRes)
```
We were using `ElementsMatch()` to avoid flakyness if results are not sorted as the expected value. Is it different now?

gouthamve replied:
Yes its different now. We return sorted label values now.

pracucci commented on pkg/storegateway/bucket_stores.go:
```diff
@@ -247,6 +247,42 @@ func (u *BucketStores) Series(req *storepb.SeriesRequest, srv storepb.Store_Seri
+
+// LabelValues implements the Storegateway proto service.
+func (u *BucketStores) LabelValues(ctx context.Context, req *storepb.LabelValuesRequest) (*storepb.LabelValuesResponse, error) {
+	spanLog, spanCtx := spanlogger.New(ctx, "BucketStores.LabelNames")
```
```suggestion
	spanLog, spanCtx := spanlogger.New(ctx, "BucketStores.LabelValues")
```

pracucci commented on pkg/storegateway/bucket_stores.go:
```diff
@@ -247,6 +247,42 @@ func (u *BucketStores) Series(req *storepb.SeriesRequest, srv storepb.Store_Seri
+
+	store := u.getStore(userID)
+	if store == nil {
+		return nil, nil
```
The caller code (in the querier expects either an error or a `storepb.LabelNamesResponse`, otherwise it panics (I think). Why don't we return an empty response here?

pracucci commented on pkg/storegateway/bucket_stores.go:
```diff
@@ -247,6 +247,42 @@ func (u *BucketStores) Series(req *storepb.SeriesRequest, srv storepb.Store_Seri
+
+	store := u.getStore(userID)
+	if store == nil {
+		return nil, nil
```
The caller code (in the querier expects either an error or a storepb.LabelValuesResponse, otherwise it panics (I think). Why don't we return an empty response here?

pracucci commented on pkg/querier/blocks_store_queryable.go:
```diff
@@ -292,14 +294,74 @@ func (q *blocksStoreQuerier) Select(_ bool, sp *storage.SelectHints, matchers ..
-func (q *blocksStoreQuerier) LabelValues(name string) ([]string, storage.Warnings, error) {
-	// Cortex doesn't use this. It will ask ingesters for metadata.
-	return nil, nil, errors.New("not implemented")
+func (q *blocksStoreQuerier) LabelNames() ([]string, storage.Warnings, error) {
```
I would like to see unit tests for this. You could get inspired by `Select()` unit tests.

pracucci commented on pkg/querier/blocks_store_queryable.go:
```diff
@@ -292,14 +294,74 @@ func (q *blocksStoreQuerier) Select(_ bool, sp *storage.SelectHints, matchers ..
-func (q *blocksStoreQuerier) LabelNames() ([]string, storage.Warnings, error) {
-	// Cortex doesn't use this. It will ask ingesters for metadata.
-	return nil, nil, errors.New("not implemented")
+func (q *blocksStoreQuerier) LabelValues(name string) ([]string, storage.Warnings, error) {
```
I would like to see unit tests for this. You could get inspired by `Select()` unit tests.

pracucci commented on pkg/querier/querier.go:
```diff
@@ -373,11 +377,96 @@ func (q querier) Select(_ bool, sp *storage.SelectHints, matchers ...*labels.Mat
+		return q.queriers[0].LabelValues(name)
+	}
+
+	// Using an errgroup here instead of channels, etc because this
+	// is a better model imo and we should move to this everywhere.
```
This looks more part of the PR description than a code comment. I would remove it.

pracucci commented on pkg/querier/querier.go:
```diff
@@ -373,11 +377,96 @@ func (q querier) Select(_ bool, sp *storage.SelectHints, matchers ...*labels.Mat
+			}
+
+			// We need values to be sorted we can merge them.
+			sort.Strings(myValues)
```
May you confirm store-gateway and "distributor" don't return it already sorted, right? If store-gateway does, then we could add it to the API spec (response is guaranteed to be sorted) and we could just sort it in the "distributor".

_Same comment applies to `LabelNames()`_

gouthamve replied:
I have this change made in the distributor, but but the querier interface we rely on is the Prometheus interface which doesn't guarantee sorted-ness.

We can always say in Cortex we always return sorted, but not sure if it will bite us sometime in the future.

pracucci commented on pkg/querier/querier.go:
```diff
@@ -373,11 +377,96 @@ func (q querier) Select(_ bool, sp *storage.SelectHints, matchers ...*labels.Mat
+		return q.queriers[0].LabelNames()
+	}
+
+	// Using an errgroup here instead of channels, etc because this
+	// is a better model imo and we should move to this everywhere.
```
No need for this comment here either.

pracucci commented on pkg/querier/blocks_store_queryable.go:
```diff
@@ -292,14 +294,74 @@ func (q *blocksStoreQuerier) Select(_ bool, sp *storage.SelectHints, matchers ..
+		return nil, nil, err
+	}
+
+	return strutil.MergeSlices(resNameSets...), resWarnings, nil
```
Each set in `resNameSets` must be already sorted to work correctly with `strutil.MergeSlices()`. Are they?

_Same comment applies to `LabelValues()`_

pracucci commented on pkg/querier/blocks_store_queryable.go:
```diff
@@ -584,6 +646,160 @@ func (q *blocksStoreQuerier) fetchSeriesFromStores(
+		reqCtx        = grpc_metadata.AppendToOutgoingContext(ctx, cortex_tsdb.TenantIDExternalLabel, q.userID)
+		g, gCtx       = errgroup.WithContext(reqCtx)
+		mtx           = sync.Mutex{}
+		nameSets      = [][]string{}
```
I think this should be called `valueSets`

pracucci commented on pkg/querier/blocks_store_queryable.go:
```diff
@@ -584,6 +646,160 @@ func (q *blocksStoreQuerier) fetchSeriesFromStores(
+		g.Go(func() error {
+			req, err := createLabelValuesRequest(minT, maxT, name, blockIDs)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create label names request")
```
```suggestion
				return errors.Wrapf(err, "failed to create label values request")
```

pracucci commented on pkg/querier/blocks_store_queryable.go:
```diff
@@ -584,6 +646,160 @@ func (q *blocksStoreQuerier) fetchSeriesFromStores(
+			if valuesResp.Hints != nil {
+				hints := hintspb.LabelValuesResponseHints{}
+				if err := types.UnmarshalAny(valuesResp.Hints, &hints); err != nil {
+					return errors.Wrapf(err, "failed to unmarshal label names hints from %s", c)
```
```suggestion
					return errors.Wrapf(err, "failed to unmarshal label values hints from %s", c)
```

pracucci commented on pkg/querier/blocks_store_queryable.go:
```diff
@@ -584,6 +646,160 @@ func (q *blocksStoreQuerier) fetchSeriesFromStores(
+				myQueriedBlocks = ids
+			}
+
+			level.Debug(spanLog).Log("msg", "received label names from store-gateway",
```
```suggestion
			level.Debug(spanLog).Log("msg", "received label values from store-gateway",
```

pracucci commented on pkg/querier/blocks_store_queryable.go:
```diff
@@ -584,6 +646,160 @@ func (q *blocksStoreQuerier) fetchSeriesFromStores(
+
+			level.Debug(spanLog).Log("msg", "received label names from store-gateway",
+				"instance", c,
+				"num labels", len(valuesResp.Values),
```
```suggestion
				"num values", len(valuesResp.Values),
```

pracucci commented on pkg/querier/blocks_store_queryable.go:
```diff
@@ -611,6 +827,69 @@ func createSeriesRequest(minT, maxT int64, matchers []storepb.LabelMatcher, skip
+		End:   maxT,
+	}
+
+	if len(blockIDs) == 0 {
```
Is this a real case? When does it happen?

gouthamve replied:
When Thanos querier queries our storegateways. I know it's a rare-case (maybe even non-existent) and can remove it if you'd like.

pracucci commented on pkg/querier/blocks_store_queryable.go:
```diff
@@ -611,6 +827,69 @@ func createSeriesRequest(minT, maxT int64, matchers []storepb.LabelMatcher, skip
+
+	anyHints, err := types.MarshalAny(hints)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to marshal request hints")
```
```suggestion
		return nil, errors.Wrapf(err, "failed to marshal label names request hints")
```

pracucci commented on pkg/querier/blocks_store_queryable.go:
```diff
@@ -611,6 +827,69 @@ func createSeriesRequest(minT, maxT int64, matchers []storepb.LabelMatcher, skip
+
+	anyHints, err := types.MarshalAny(hints)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to marshal request hints")
```
```suggestion
		return nil, errors.Wrapf(err, "failed to marshal label values request hints")
```

pracucci reviewed (commented):
Good job! This work is pretty solid 👏  I left some comments. Most are minor thing and small improvements, but there's a comment about the ordering requirement of `strutil.MergeSlices()` which I would like you to investigate and also I would be glad to see some unit tests on blocks store queryable. Thanks! 🚀

gouthamve commented:
@pracucci Thanks for the detailed review! I fixed all the suggestions and added the tests. This is good for another round of review. I'll dig into and add comments about the sortedness of names and values shortly.

gouthamve commented on pkg/querier/blocks_store_queryable_test.go:
```diff
@@ -544,6 +545,474 @@ func TestBlocksStoreQuerier_Select(t *testing.T) {
+		t.Run(testName, func(t *testing.T) {
+			// Splitting it because we need a new registry for names and values.
+			// And also the initial expectedErr checking needs to be done for both.
+			for _, testFunc := range []string{"LabelNames", "LabelValues"} {
```
I feel like there should be a saner approach than this but my brain can't come up with any. Any ideas @pracucci / @pstibrany?

pracucci replied:
I think you did great. I would have probably done the same. We have other tests where we adopt this approach. LGTM.

pracucci reviewed (commented):
Thanks @gouthamve for addressing my feedback! LGTM (modulo a couple of nits). Looking forward to your comment about sorting and then we can go 🚀

pstibrany reviewed (commented):
Sorry for late review. Overall it looks good to me (with Marco's comments fixed). I'd like to see less duplicated code, if possible, but that can also be addressed later.

gouthamve commented:
@pracucci I've addressed all your comments, but @pstibrany I kinda kept the suggestion of deduplicating code for another PR. I'll soon open a follow up PR for this once I have some bandwidth.

pracucci reviewed (approved):
LGTM, thanks!
